types/decimal.go

```diff
@@ -1,28 +1,91 @@
-// more than four digits. In Go this is stored as an int64, the precision is
-// defined by the constant DecimalPrecision.
+// more than four digits. A decimal value can range from -922337203685477.5808
+// to 922337203685477.5807.
 type Decimal struct {
-	Value int64
+	value int64
+}
+
+// NewDecimal returns a Decimal value of i * 10^exponent.
+func NewDecimal(i int64, exponent int) (Decimal, error) {
+	if exponent < -4 || exponent > 14 {
+		return Decimal{}, fmt.Errorf("%w: exponent value of %v exceeds maximum range of Decimal", ErrDecimal, exponent)
+	}
+
+	var intPart int64
+	var fracPart int64
+	if exponent <= 0 {
+		intPart = i / int64(math.Pow10(-exponent))
+		fracPart = i % int64(math.Pow10(-exponent)) * int64(math.Pow10(4+exponent))
+	} else {
+		intPart = i * int64(math.Pow10(exponent))
+		if i > 0 && intPart < i {
+			return Decimal{}, fmt.Errorf("%w: value %ve%v would overflow", ErrDecimal, i, exponent)
+		} else if i < 0 && intPart > i {
+			return Decimal{}, fmt.Errorf("%w: value %ve%v would underflow", ErrDecimal, i, exponent)
+		}
+	}
+
+	if intPart > 922337203685477 || (intPart == 922337203685477 && fracPart > 5807) {
+		return Decimal{}, fmt.Errorf("%w: value %ve%v would overflow", ErrDecimal, i, exponent)
+	} else if intPart < -922337203685477 || (intPart == -922337203685477 && fracPart < -5808) {
+		return Decimal{}, fmt.Errorf("%w: value %ve%v would underflow", ErrDecimal, i, exponent)
+	}
+
+	return Decimal{value: intPart*decimalPrecision + fracPart}, nil
 }
 
-// UnsafeDecimal creates a decimal via unsafe conversion from int, int64, float64.
-// Precision may be lost and overflows may occur.
-func UnsafeDecimal[T int | int64 | float64](v T) Decimal {
-	return Decimal{Value: int64(v * DecimalPrecision)}
+// NewDecimalFromInt returns a Decimal with the whole integer value provided
+func NewDecimalFromInt[T constraints.Signed](i T) (Decimal, error) {
+	return NewDecimal(int64(i), 0)
 }
 
-// DecimalPrecision is the precision of a Decimal.
-const DecimalPrecision = 10000
+// NewDecimalFromFloat returns a Decimal that approximates the given floating point value.
+// The value of the Decimal is calculated by multiplying it by 10^4, truncating it to
+// an int64 representation to cut off any digits beyond the four allowed, and passing it
+// as an integer to NewDecimal() with -4 as the exponent.
+//
+// WARNING: decimal representations of more than 6 significant digits for float32s and 15
+// significant digits for float64s can be lossy in terms of precision. To create a precise
+// Decimal above those sizes, use the NewDecimal constructor.
+func NewDecimalFromFloat[T constraints.Float](f T) (Decimal, error) {
+	f = f * decimalPrecision
+	if f > math.MaxInt64 {
+		return Decimal{}, fmt.Errorf("%w: value %v would overflow", ErrDecimal, f)
+	} else if f < math.MinInt64 {
+		return Decimal{}, fmt.Errorf("%w: value %v would underflow", ErrDecimal, f)
+	}
+
+	return NewDecimal(int64(f), -4)
```

Comment:
You've already munged the number to the correct int64. Doesn't seem necessary to do the work again by calling `NewDecimal` here?

Reply:
Nice. I'll make the change. I generally like to err on the side of going through the "one true" constructor for types, but it does seem kind of silly in this case.